Xml/dz02

AskaV opened the pull request:
Homework 2_ RecyclerView - Contacts List, Interactivity & Custom Google Reg

• Implemented contacts list screen: XML RecyclerView, Adapter/ViewHolder, ViewModel data handling, avatar loading with rounded shapes.

• Added interactivity: delete contact (click/swipe) with undo, add contact via dialog

• Created custom Google registration button with configurable XML attributes.

AskaV committed:
Initial project setup

(cherry picked from commit 96ffb7d5f900693e27cfc30a7c61d3a4b0f8f987)

AskaV committed:
added my profile markup

(cherry picked from commit 587de8a1ac5e112ba53d1d98ab8ca8bfa63f13a6)

AskaV committed:
Fixed based on review comments for the my profile markup task

The xml file code has been corrected - the numeric value has been moved to a separate file.
The icon and description for 1 social network have been changed
Also, the task interface 1 in Compose format has been added
The main file has been corrected - the basic values ​​have been removed and the code for launching and checking the Compose interface of task 1 has been added

(cherry picked from commit a776b8cfe6374540282710825bb0705bcef0ece1)

AskaV committed:
adding changes to task 1 part 2

Added values ​​to the file with sizes + arranged and logically arranged them
Changed the name of the xml file to a more appropriate one
Added colors for the theme
Removed the nested component in the xml file
Added a screen config for previewing work with compose
Fixed the numerical and color values ​​in the kotlin files by replacing them with the appropriate values

(cherry picked from commit 2a5ad9889030330ed8c0bb22f17ffb7f6920a300)

AskaV committed:
adding changes to task 1 part 3

Removed additional parameters from functions and reworked them in code
Removed dimens as parameters for Preview in order to avoid duplication of code in subsequent files, and changed the name so that it was possible to add other screens and modify them in the configuration
I removed the numbers in the style values ​​that I used for convenience, replacing them with more understandable names
More minor edits

(cherry picked from commit 77aea6c9d59978ca66dfd0c6aafb66c8af04ed51)

AskaV committed:
adding changes to task 1 part 4

Changed the text View my contacts to be more readable, and added .uppercase() to the code to match the template
added the color parameter for better code readability

(cherry picked from commit 7f309d2097b407af76d91314e28e6576ad2c7e47)

AskaV committed:
added login page, sign up page markup

AskaV committed:
Changing project settings and removing unnecessary files

AskaV committed:
Changing project settings and removing unnecessary files

AskaV committed:
added xml auth activity

Changes have been made to the manifest regarding the start page
Added changes for class for selecting the launch of activities as a start (xml or compose version)
added kotlin for xml versions of 2 screens
Added colors, line sizes and values ​​for the implementation of added screens
- added email-password validation
- added navigation from auth to main (with data passing)

AskaV committed:
added theme layout for Xml screens

AskaV committed:
added contacts recycler view xml version

AskaV committed:
implemented data initialization via viewmodel xml version

implemented data initialization via viewmodel xml version
small visual fixes

AskaV committed:
added image load via url

AskaV committed:
implemented content provider for phone contacts

AskaV committed:
implemented content provider for phone contacts

AskaV committed:
implemented contact removal, added  undo snackbar for xml

AskaV committed:
correction of condition

AskaV committed:
implemented add contact dialog for xml v

AskaV committed:
Revert "Changing project settings and removing unnecessary files"

This reverts commit f1604568c446f97124714088e4cf1d9e60c9ffea.

AskaV committed:
dz01(xml): added auth activity

AskaV committed:
dz01(compose): commit compose part

AskaV committed:
dz01(compose): add auth activity commit file

AskaV committed:
add dimen file

AskaV committed:
add ness files for working dz02

kogorbach commented on app/src/main/java/com/spp/android/myapplication/data/UserPreferences.kt:
```diff
@@ -0,0 +1,27 @@
+
+object UserPreferences {
+
+    private val EMAIL_KEY = stringPreferencesKey("email")
```
I would extract 'magic' strings "email" and "user_prefs" to constants

kogorbach commented on app/src/main/java/com/spp/android/myapplication/data/UserPreferences.kt:
```diff
@@ -0,0 +1,27 @@
+
+    private val EMAIL_KEY = stringPreferencesKey("email")
+
+    suspend fun saveEmail(context: Context, email: String) {
```
usually a bad practice to pass context in parameters, but it will be complicated to fix here, what you can do is inject this class in a repository, since Datastore is basically a data source, and the class responsible for data sources is a repository
TLDR:
- declare a singleton of this class in your di folder, I think you're already familiar with Hilt
- inject it in your repository as a data source

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/BaseActivity.kt:
```diff
@@ -0,0 +1,20 @@
+import androidx.appcompat.app.AppCompatActivity
+import com.spp.android.myapplication.R
+
+open class BaseActivity : AppCompatActivity() {
```
Like how you use a base activity, it's an arguable approach but at this stage I think it's good you used it

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/BaseActivity.kt:
```diff
@@ -0,0 +1,20 @@
+open class BaseActivity : AppCompatActivity() {
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        val prefs: SharedPreferences = getSharedPreferences("user_prefs", MODE_PRIVATE)
```
Kudos for learning both SharedPreferences and Datastore, but in the future better stick to one, probably DataStore as a more modern tool

AskaV replied:
Thanks for the remark! For now I’ll keep both approaches, but in the next version (task 5) I plan to clean up the code and fully migrate to DataStore.

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/LoginActivityXml.kt:
```diff
@@ -0,0 +1,90 @@
+                setContentView(R.layout.login_page)
+                setupViews()
+
+                if (savedInstanceState != null) {
```
try managing saved state through viewmodel instead of activity

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/LoginActivityXml.kt:
```diff
@@ -0,0 +1,90 @@
+                if (savedInstanceState != null) {
+                    val email = savedInstanceState.getString("email_text", "")
+                    val password = savedInstanceState.getString("password_text", "")
+                    findViewById<EditText>(R.id.editTextTextEmailAddress).setText(email)
```
please add viewbinding and use it instead of findViewById

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/ValidationUtils.kt:
```diff
@@ -0,0 +1,46 @@
+
+object ValidationUtils {
+
+    fun validateEmailAndPassword(
```
it's a util class, so it shouldn't be aware of the UI layer

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/LoginActivityXml.kt:
```diff
@@ -0,0 +1,90 @@
+
+    private fun navigateToMain(email: String) {
+        val intent = Intent(this, MainActivityXml::class.java).apply {
+            putExtra("email", email)
```
another magic string, they are quite dangerous because you can fail a transition with a simple mistype

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/MainActivityXml.kt:
```diff
@@ -0,0 +1,40 @@
+        userNameTextView.text = userName
+
+        findViewById<Button>(R.id.editProfileBtn).setOnClickListener {
+            Toast.makeText(this, "Edit Profile clicked", Toast.LENGTH_SHORT).show()
```
not really a mistake, but notice how you repeat the boilerplate code for a toast, consider creating an extension function for it

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/MainActivityXml.kt:
```diff
@@ -0,0 +1,40 @@
+        }
+    }
+
+    private fun parseNameFromEmail(email: String): String {
```
ideally it should be in a separate class, maybe a util class or a parser, because activity is a UI class, and in this case you would need to change it if the parsing logic changes (S from SOLID)

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/SignUpActivityXml.kt:
```diff
@@ -0,0 +1,56 @@
+            )
+
+            if (allValid) {
+                lifecycleScope.launch {
```
why do you need a coroutine here?

AskaV replied:
Because of the saveEmail call and saving data to the DataStore, this block can only be written this way, otherwise there will be an error.

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+class ContactAdapter(
+    private val contacts: MutableList<Contact>,
+    private val onDeleteClick: (Contact, Int) -> Unit
+) : RecyclerView.Adapter<ContactAdapter.ContactViewHolder>() {
```
best use ListAdapter instead of RecyclerView.Adapter, it forces DiffUtils and eases their integration

AskaV replied:
Thanks! Agreed — ListAdapter + DiffUtil is the right approach. I’ll keep RecyclerView.Adapter in this PR to limit scope and migrate on it in Task 5

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+
+    inner class ContactViewHolder(view: View) : RecyclerView.ViewHolder(view) {
+        val avatarImageView: ImageView = itemView.findViewById(R.id.avatarImageView)
+        val nameTextView: TextView = itemView.findViewById(R.id.nameTextView)
```
quick reminder of using viewbinding instead of findViewById, needed here as well

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactAdapter.kt:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    override fun getItemCount(): Int = contacts.size
+
+    fun removeContactAt(position: Int) {
+        contacts.removeAt(position)
+        notifyItemRemoved(position)
+    }
+
+    fun restoreContact(contact: Contact, position: Int) {
+        contacts.add(position, contact)
+        notifyItemInserted(position)
+    }
+
```
If ListAdapter is implemented correctly, these won't be necessary, you'll just update the dataset and the adapter will handle it for you

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactsActivityXml.kt:
```diff
@@ -0,0 +1,204 @@
+        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
+        recyclerView.layoutManager = LinearLayoutManager(this)
+
+        viewModel = ViewModelProvider(this)[ContactsViewModel::class.java]
```
you can inject viewmodel with a simpler byViewmodels delegate, and Hilt for its dependencies

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactsActivityXml.kt:
```diff
@@ -0,0 +1,204 @@
+        viewModel.contacts.observe(this) { contactList ->
+            val mutableList = contactList.toMutableList()
+
+            adapter = ContactAdapter(mutableList) { contactToDelete, position ->
```
I'd suggest using a named parameter for lambda here, at this point it's not obvious it's deleting the contact, I had to scroll back to adapter to know what it was

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactsActivityXml.kt:
```diff
@@ -0,0 +1,204 @@
+        grantResults: IntArray
+    ) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if (requestCode == 100 && grantResults.firstOrNull() == PackageManager.PERMISSION_GRANTED) {
```
not sure but I assume 100 is a magic number here?

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactsActivityXml.kt:
```diff
@@ -0,0 +1,204 @@
+        }
+    }
+
+    private fun loadContactsFromPhone() {
```
ideally should be handled in a separate class (or object)

AskaV replied:
Thanks for the feedback! For now I’ll keep the logic inside Activity, but in the next task (Task 5) I plan to refactor it.

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/contacts/ContactsActivityXml.kt:
```diff
@@ -0,0 +1,204 @@
+                    val newContact = Contact(
+                        name = name,
+                        position = position,
+                        avatarUrl = "https://i.pravatar.cc/150?img=${(1..70).random()}"
```
I think you already have this random fetch in your viewmodel?

kogorbach commented on app/src/main/java/com/spp/android/myapplication/xmlscreens/util/extensions/ImageViewExtensions.kt:
```diff
@@ -0,0 +1,53 @@
+import com.spp.android.myapplication.R
+import com.squareup.picasso.Picasso
+
+class ImageViewExtensions {
```
nice work on managing three image load libraries here, I think it can be an object instead of a class though, thus forcing a singleton

AskaV committed:
Code fixes after review for xml task2

- extract 'magic' strings "email" and "user_prefs" to constants
- implement viewbinding instead of findViewById
- refactoring: replace ViewModelProvider with delegate using viewModels()
- remove ValidationUtils class to utils layer
- moving magic strings to string resources
- extract toast boilerplate into extension
- convert ImageViewExtensions from class to object (singleton)
- reuse generateAvatarUrl() from ViewModel instead of duplicating logic
- replace magic number 100 with REQUEST_CODE_CONTACTS_PERMISSION constant
- change name onDeleteClick to onDeleteContact
- extract parseNameFromEmail to ValidationUtils and use it in MainActivityXml
- move saved state handling from Activity to LoginViewModel
